### ISSUE TITLE
Use with pam_mount

### ISSUE THREAD

**Issue by aacebedo** [shimunn/fido2luks#20]
I successfully created a luks container and mounted it with fido2luks (great tool BTW !)
I would like to integrate it with pam_mount and I don't know where to begin.

My current setup dynamically opens a luks container containing my home when I logged in through lightdm.
The password of my session and of my container are the same so I enter the password just one time, it is used by lightdm and passed to cryptsetup to open the container.

I would like to use my U2F key to mount the luks container. As I use a solokey it requests a pin code at each time so I guess I would have to enter it during the lightdm login after the password. Or I can change the password to be the pincode to reuse it.

Do you have an idea of how I can use fido2luks to do this ?

**Comment by shimunn** [comment]
I haven't tried anything like that myself since I prefer to encrypt my whole system but I think it may be possible to wrap fido2luks in a shell script which can then be called by pam_mount:

pam_mount.conf
```
<cryptmount>fido2luks-wrapper.sh ...</cryptmount>
<cryptumount>umount.crypt %(MNTPT)</cryptumount>
```
fido2luks-wrapper.sh(I not exactly sure if pam_mount actually passes the password via stdin but that should be easy to figure out):
```
set -a
. /etc/fido2luks.conf
fido2luks open $1 "device_name" --salt 'keyfile:/dev/stdin'
```

This is what just came to mind and should work after some tweaking but if there is interest in a pam_module I might be able to put one together as well.

Passing the PIN is already painful and I do not have a neat solution on hand, you could either just hardcode it in your script or.

**Comment by aacebedo** [comment]
Thanks for the quick answer.
I tried to pass the pincode using a.simple script to test it without success.
It ses that the cli input after the AUTHENYICATOR PIN message doesn't come from fido2luks. 
Do you know how to pass it in hardcoded way ? Or better to disable the pin on a solokeys?

**Comment by shimunn** [comment]
> I tried to pass the pincode using a simple script to test it but without success.

By default it reads the pin from tty which isn't available in the pam_mount script context, you can however use `--pin-source <file>` to read the PIN from a file instead.

> Do you know how to pass it in hardcoded way 

Obviously passing it via file isn't ideal and will likely be changed in the next major version but in the meantime you could try to pass the PIN via stdin and the password as string.

```

set -a
. /etc/fido2luks.conf
read PASSWORD
echo "mysecret pin" | fido2luks open $1 "device_name" --salt 'string:$PASSWORD' --pin --pin-source /dev/stdin
```

> Or better to disable the pin on a solokeys?

Unfortunately there is none apart from resetting key and losing access to all services you paired before.

**Comment by aacebedo** [comment]
Unfortunately, getting the pincode from the pipe or a file returns the following error:
AuthenticatorError { cause: FidoError(FidoError(

Device returned error: CborError: 0x31: PIN Invalid.)

Error while decoding CBOR from device.) }

**Comment by shimunn** [comment]
Sorry my bad, you have to use printf instead of echo since echo will include a newline

**Comment by aacebedo** [comment]
Thank you ! That's working !
I can finally fully use my solokey !
I wrote a small script for the mount/unmount and it works flawlessly.

**Comment by madonius** [comment]
@shimunn if I see it correctly this issue was resolved by #21, no?
